text/chapter10.md

```diff
@@ -391,416 +437,935 @@ foreign import isUndefined :: forall a. Undefined a -> Boolean
+```text
+$ spago repl
 
-```haskell
-foreign import data Effect :: Type -> Type
+> import Test.Examples
+> import Data.Tuple
+> boldConstraint (Tuple 1 "Hat")
+"(TUPLE 1 \"HAT\")!!!"
 ```
 
-As a simple example, consider the `random` function defined in the `random` package. Recall that its type was:
+Here's another example demonstrating multiple constraints:
 
-```haskell
-foreign import random :: Effect Number
+```js
+exports.showEquality = Eq => Show => a => b => {
+  if (Eq.eq(a)(b)) {
+    return "Equivalent";
+  } else {
+    return Show.show(a) + " is not equal to " + Show.show(b);
+  }
+}
 ```
 
-The definition of the `random` function is given here:
+```hs
+foreign import showEquality :: forall a. Eq a => Show a => a -> a -> String
+```
 
-```javascript
-exports.random = Math.random;
+```text
+$ spago repl
+
+> import Test.Examples
+> import Data.Maybe
+> showEquality Nothing (Just 5)
+"Nothing is not equal to (Just 5)"
 ```
 
-Notice that the `random` function is represented at runtime as a function of no arguments. It performs the side effect of generating a random number, and returns it, and the return value matches the runtime representation of the `Number` type: it is a non-null JavaScript number.
+## Effectful Functions
 
-As a slightly more interesting example, consider the `log` function defined by the `Effect.Console` module in the `console` package. The `log` function has the following type:
+Let's extend our `boldConstraint` function to log to the console. Logging is an `Effect`, and `Effect`s are represented in JavaScript as a function of zero arguments, `()` with arrow notation:
 
-```haskell
-foreign import log :: String -> Effect Unit
+```js
+exports.yell = Show => x => () =>
+  console.log(Show.show(x).toUpperCase() + "!!!");
 ```
 
-And here is its definition:
+The new foreign import is the same as before, except that the return type changed from `String` to `Effect Unit`.
+```hs
+foreign import yell :: forall a. Show a => a -> Effect Unit
+```
 
-```javascript
-exports.log = function (s) {
-  return function () {
-    console.log(s);
-  };
+When testing this in the repl, notice that string is printed directly to the console (instead of being quoted) and a `unit` value is returned.
+
+```text
+$ spago repl
+
+> import Test.Examples
+> import Data.Tuple
+> yell (Tuple 1 "Hat")
+(TUPLE 1 "HAT")!!!
+unit
+```
+
+There are also `EffectFn` wrappers from `Effect.Uncurried`. These are similar to the `Fn` wrappers from `Data.Function.Uncurried` that we've already seen. These wrappers let you call uncurried effectful functions in PureScript.
+
+You'd generally only use these if you want to call existing JavaScript library APIs directly, rather than wrapping those APIs in curried functions. So it doesn't make much sense to present an example of uncurried `yell`, where the JavaScript relies on PureScript type class members, since you wouldn't find that in the existing JavaScript ecosystem.
+
```

Comment:
Is it even possible to grab constraint args (like `Show`) with runFn? Maybe this should be a note in the Constraint section.